p2p/server/msgserver.go

```diff
@@ -103,29 +105,37 @@ func NewMsgServer(ctx context.Context, network Service, name string, requestLife
 
 // Close stops the MessageServer
 func (p *MessageServer) Close() {
-	p.exit <- struct{}{}
-	<-p.exit
-	p.workerCount.Wait()
+	select {
+	case <-p.exit:
+		// channel already closed
+		// (since nothing is ever sent on this channel)
+	default:
+		close(p.exit)
+		<-p.closed
```

Comment:
I know this is a draft, though I'm not sure why we need both `closed` and `exit`.. actually since context is passed in, can't we just use `context.Done` for all of that?